### PR TITLE
[docs] Migrate frankenstyle document

### PR DESCRIPTION
**Pull request by sharidas**

### CODE REVIEW

File: general/development/policies/codingstyle/frankenstyle.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
+---
+title: Frankenstyle
+tags:
+  - Plugins
+---
+'Frankenstyle component names' refers to the naming convention that is used to uniquely identify a Moodle plugin based on the type of plugin and its name.  They are used throughout the Moodle code (with a notable exception being the css class names in the themes).
+
+Martin Dougiamas invented the word 'frankenstyle' to describe this naming system which was invented by Petr Škoda.
+
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:9]: I'd suggest removing this anecdote. It's nice, but unnecessary.

```suggestion
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
+---
+title: Frankenstyle
+tags:
+  - Plugins
+---
+'Frankenstyle component names' refers to the naming convention that is used to uniquely identify a Moodle plugin based on the type of plugin and its name.  They are used throughout the Moodle code (with a notable exception being the css class names in the themes).
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:6]: I might suggest rewording this slightly:

```suggestion
The term 'Frankenstyle component names' refers to the naming convention that is used to uniquely identify a Moodle plugin based on the type of plugin and its name.  They are used throughout the Moodle code (with a notable exception being the css class names in the themes).
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
+---
+title: Frankenstyle
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:2]: I might suggest renaming the page title to 'Frankenstyle component names'

To keep the sidebar label short, we can use the `sidebar_label:` yaml attribute: https://docusaurus.io/docs/next/api/plugins/@docusaurus/plugin-content-docs#sidebar_label

So:
```suggestion
title: Frankenstyle Component Names
sidebar_label: Frankenstyle
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+## Plugin types
+
+See [Plugin types](/docs/apis/plugintypes/) page for the list of all supported plugin types in Moodle and their frankenstyle prefix.
+
+To get a definitive list in your version of Moodle 2.x, use a small Moodle script with <tt>print_object(get_plugin_types());</tt>.
+
+## Core subsystems
+
+Subsystems in Moodle are not plugins themselves but can be referred to using **core_xxx** where xxx is the subsystem name as defined in get_core_subsystems().
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:27]:
```suggestion
Subsystems in Moodle are not plugins themselves but can be referred to using **core_[subsystem.]** where the subsystem is defined in get_core_subsystems().
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+You can see these being used in the @package parameter in phpdocs, or in the [webservice function names](https://docs.moodle.org/dev/Web_services_Roadmap). Core subsystems can provide own strings via a file stored in <tt>lang/en/{`subsystemname`}.php</tt>. Some of them have a dedicated location with libraries, autoloaded classes and other resources.
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:29]: There are few more places that now use the name since we wrote these docs and they're pretty imporant so I'd suggest adding them and turning it into a list:

```suggestion
Other places that you may see these being used include:

- the PHP Namespace used to autoload classes
- the prefix used in JavaScript module names
- the prefix used in template names
- the `@package` parameter in phpdocs
- the [webservice function names](https://docs.moodle.org/dev/Web_services_Roadmap)

Core subsystems can provide own strings via a file stored in `lang/en/{`subsystemname`}.php`. Some of them have a dedicated location with libraries, autoloaded classes and other resources.
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+| User key | core_userkey |  |
+| Web service | core_webservice | /webservice |
+
+## Usage
+
+Frankenstyle component names are used in:
+
+### Function names
+
+All plugin functions must start with full frankenstyle prefix. For backwards compatibility modules may also use **modulename_** as prefix.
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:107]: We no longer recommend the use of global functions so I'd suggest adding a warning admonition:

```suggestion
All plugin functions must start with full frankenstyle prefix.

:::note Activity modules

For backwards compatibility modules may also use `modulename_` as prefix.

:::

:::warning

Something about global functions not being recommended. Please use an autoloaded class.

:::


```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+### Class names
+
+- All the component classes must sit under the classes ([auto-loaded](https://docs.moodle.org/dev/Automatic_class_loading)) directory, belonging to their frankenstyle namespace and having a natural name, ex: \mod_forum\example.
+- The use of class names using the frankenstyle prefix is now deprecated (see [Coding style](./index.md#namespaces)), ex: mod_forum_example.
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:112]: Convert this line to an admonition:

```suggestion

:::warning Non-namespaced classes

The use of non-namespaced classes using only the frankenstyle prefix is now deprecated.

 See [Coding style](./index.md#namespaces) for more information.
 
 For example, the class. name `mod_forum_example` should be written as `mod_forum\example`.

:::
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+Frankenstyle component names are used in:
+
+### Function names
+
+All plugin functions must start with full frankenstyle prefix. For backwards compatibility modules may also use **modulename_** as prefix.
+
+### Class names
+
+- All the component classes must sit under the classes ([auto-loaded](https://docs.moodle.org/dev/Automatic_class_loading)) directory, belonging to their frankenstyle namespace and having a natural name, ex: \mod_forum\example.
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:111]: Don't use `ex:`. This may not be easy to understand for some readers who are not native English speakers.

```suggestion
All component classes must be placed under the classes directory, which allows them to be ([autoloaded](https://docs.moodle.org/dev/Automatic_class_loading)). These should be placed in a namespace according to their frankenstyle component name, and having a natural class name, for example a discussion class in the forum activity should be in the `mod_forum` namespace and may have a class name of `dicussion` - `\mod_forum\discussion`.

```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+- The use of class names using the frankenstyle prefix is now deprecated (see [Coding style](./index.md#namespaces)), ex: mod_forum_example.
+
+### Constants
+
+All plugin constants must start with uppercase frankenstyle prefix, example MOD_FORUM_XXXX.
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:116]:
```suggestion
All plugin constants must start with uppercase frankenstyle prefix, for example `MOD_FORUM_XXXX`.

:::note

The use of constants is not recommended and, where possible, a class constant on an autoloaded class should be used.

This allows uses of the constant to autoload the content without needing to manually require any files.

:::
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+### Table names
+
+All table names for a plugin must begin with its frankenstyle name (after the standard Moodle table prefix).  
+
+(The exception to this rule is Moodle activities which (for historical reasons) do not have mod_ in front of the plugin name)
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:122]:
```suggestion

:::warning Activity modules exception

The exception to this rule is Moodle activities which (for historical reasons) do not have plugin type (`mod_`) as a prefix to the plugin name.

:::
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+Examples: mdl\_**local_`coolreport`**, mdl\_**local_`coolreport`**_users
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:124]: Maybe turn into a list:

```suggestion
Examples: 

- `local_coolreport`
- `local_coolreport_users`
- `forum` - for the mod_forum component
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+### Plugin configuration table
+
+In the **config_plugins** table, column **plugin**, the frankenstyle name is used.
+
+### Capabilities
+
+All capabilities for a plugin use the frankenstyle name, except with a / instead of a _.
+
+Example: **mod/quiz**:`viewattempt`
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:134]:
```suggestion
For example:

- `mod/quiz:viewattempt`
- `block/library:readbook`
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+
+### Language files
+
+The main language file for each plugin (with the notable exception of activity modules) is the frankenstyle component name.
+
+Examples:  /blocks/participants/lang/en/**block_participants**.php
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:140]: You get the drill - try to give multiple examples and includea ny exceptions (liek activity modules in this case).

Again, the exceptions (mod_) should be in an admonition like a `:::warning`

File: general/development/policies/codingstyle/frankenstyle.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+### Module Subplugins
+
+It is possible to extend modules without having to change the basic module's code. See [Subplugins](https://docs.moodle.org/dev/Subplugins) for details.
+
+### Other places (TODO)
+
+- @package declarations in phpdocs, see [Coding style#PHPDoc](./index.md#phpdoc)
+- [web service function names](https://docs.moodle.org/dev/Web_services_Roadmap)
+- [Moodle Plugins database](http://moodle.org)
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:152]: Need to either document or add these to the list:

- JS module names
- template names

File: general/development/policies/codingstyle/frankenstyle.md [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+Please add more as they come up.
+
+## Naming theme variants
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:156]:
```suggestion
## Theme name variants
```

File: general/development/policies/codingstyle/frankenstyle.md [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+Themes that are principally derivatives of some other theme, should be named in a way that clearly implies they are variants, not upgrades, of the parent theme. The format of the name should follow and extend the standard component naming format. So for example themes based on Boost, if they want to mention the word "boost" in the component name, should be named like
```
Comment (andrewnicols) [inline general/development/policies/codingstyle/frankenstyle.md:158]: Wow - this one is terrible advice! What if they want to switch to a different parent theme!? We even recommend against this in the next example! Scrap it.

```suggestion
Themes are typically a derivatives of some other theme. Where this is the case, you should _avoid_ includingt he parent theme name in your theme's name.
```

**Review by andrewnicols — CHANGES_REQUESTED** [review]
Seems like a lot of changes but it's really not. The original doc is wwwaaaaay out of date and missing key parts, and we should try and use more consistent examples and wording.

**Comment by sharidas** [comment]
Hi,
I have addressed the changes requested. Let me know if this looks ok.
Thanks,
Sujith

**Review by andrewnicols — APPROVED** [review]
